### PR TITLE
Defer keyframe resolution

### PR DESCRIPTION
**Pull request by mattgperry**
When keyframes aren't inherently mixable, we make them mixable by measuring the DOM or resolving values from it.

```javscript
animate(element, {
  height: "auto", // What is auto?
  width: ["50%", "100px"], // How do these mix?
  x: [null, 100], // What is the initial keyframe?
  background: "var(--tint)" // What color is this?
})
```

This is currently done synchronously per `animate()` call / `motion` component animation which can lead to layout thrashing when many are performed and/or when mixed in amongst other layout-measuring effects.

Screenshot 2024-02-27 at 10 50 19

## Solution

This PR defers keyframe resolution until the following animation frame. All reads and writes are batched so we have at max 1 style recalculation/ layout reflow.

It also refactors JS and WAAPI animations into a `MainThreadAnimation` and `AcceleratedAnimation` class inheriting from `BaseAnimation` to share as much of this new logic as possible.

Animations still return synchronously-available animation controls like `.play()` / `.time` etc. As much of this is computed or otherwise relies on resolved keyframes, they internally use `this.resolved` which is behind a getter. This getter will synchronously flush all pending animations as a de-opt when accessed. Some methods have already been changed to allow pending operations but this can be improved further. In the absolute worst case, where animations are created/accessed

```
const a = animate()
a.time

const b = animate()
b.time
```

Performance is the same as today. But in the best case, cold start performance is 2.5x faster and unit conversion performance is 6x faster, and much of that work is lifted out of the effect and into the next animation frame.

Screenshot 2024-02-27 at 10 56 39

Bonus:
- Support unit conversion and CSS variables in keyframes

Fixes https://github.com/framer/motion/issues/2168
Fixes https://github.com/framer/motion/issues/2309
Fixes https://github.com/framer/motion/issues/2339

### CODE REVIEW

File: dev/benchmarks/cold-start-anime.html [hunk 1]
```diff
@@ -28,15 +28,15 @@
             }
```
Comment (mattgperry) [inline dev/benchmarks/cold-start-anime.html:28]: Can ignore the `/benchmarks`

File: packages/framer-motion/cypress/integration/drag-to-reorder.ts [hunk 2]
```diff
@@ -201,7 +201,7 @@ describe("Drag to reorder", () => {
                     const y = step > 0 ? delta : -delta
                     chain = chain
                         .trigger("pointermove", 360, baseY + y, { force: true })
-                        .wait(50)
+                        .wait(100)
```
Comment (mattgperry) [inline packages/framer-motion/cypress/integration/drag-to-reorder.ts:204]: Flakey test

File: packages/framer-motion/src/animation/__tests__/animate-waapi.test.ts [hunk 3]
```diff
@@ -12,6 +13,8 @@ describe("animate() with WAAPI", () => {
             { duration: 1, transform: { duration: 2 } }
         )
 
+        await nextFrame()
```
Comment (mattgperry) [inline packages/framer-motion/src/animation/__tests__/animate-waapi.test.ts:16]: WAAPI `nextFrame` calls generally added as the WAAPI animation is only initialised after keyframes are resolved. We could add more custom logic here to say if something is WAAPI we can resolve fewer keyframes as it can handle CSS vars etc itself

File: packages/framer-motion/src/animation/hooks/animation-controls.ts [hunk 4]
```diff
@@ -1,13 +1,41 @@
...
 import type { VisualElement } from "../../render/VisualElement"
-import { AnimationControls } from "../types"
+import { AnimationControls, AnimationDefinition } from "../types"
 import { animateVisualElement } from "../interfaces/visual-element"
 
 function stopAnimation(visualElement: VisualElement) {
     visualElement.values.forEach((value) => value.stop())
 }
 
+function setVariants(visualElement: VisualElement, variantLabels: string[]) {
```
Comment (mattgperry) [inline packages/framer-motion/src/animation/hooks/animation-controls.ts:11]: Moved in from `utils/setters` as only used in this file.

File: packages/framer-motion/src/animation/interfaces/motion-value.ts [hunk 5]
```diff
@@ -36,32 +39,10 @@ export const animateMotionValue = (
         let { elapsed = 0 } = transition
         elapsed = elapsed - secondsToMilliseconds(delay)
 
-        const keyframes = getKeyframes(
```
Comment (mattgperry) [inline packages/framer-motion/src/animation/interfaces/motion-value.ts:39]: Moved to `can-animate`

File: packages/framer-motion/src/render/utils/animation-state.ts [hunk 6]
```diff
@@ -212,8 +214,9 @@ export function createAnimationState(
              * Build an object of all the resolved values. We'll use this in the subsequent
              * animateChanges calls to determine whether a value has changed.
              */
+            // TODO Resolve with options
```
Comment (adamseckel) [inline packages/framer-motion/src/render/utils/animation-state.ts:217]: Does this TODO need addressing before merge? Or is it no longer relevant?

File: packages/framer-motion/src/render/svg/SVGVisualElement.ts [hunk 7]
```diff
@@ -33,6 +33,7 @@ export class SVGVisualElement extends DOMVisualElement<
     }
 
     readValueFromInstance(instance: SVGElement, key: string) {
+        // console.log("reading", key, "from", instance)
```
Comment (adamseckel) [inline packages/framer-motion/src/render/svg/SVGVisualElement.ts:36]:
```suggestion
```

File: packages/framer-motion/src/render/html/utils/make-none-animatable.ts [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+    if (animatableTemplate && name) {
+        for (const noneIndex of noneKeyframeIndexes) {
+            unresolvedKeyframes[noneIndex] = getAnimatableNone(
+                name,
+                animatableTemplate
+            )
+        }
+    }
+}
```
Comment (adamseckel) [inline packages/framer-motion/src/render/html/utils/make-none-animatable.ts:33]: I'm having a hard time understanding what this is doing. What is the template, just some keyframe value like a color string or numeric value?

File: packages/framer-motion/src/render/utils/animation-state.ts (hunk 6 above)
Comment (mattgperry) [reply packages/framer-motion/src/render/utils/animation-state.ts:217]: Removed, I think this was an idle thought

File: packages/framer-motion/src/gestures/hover.ts [hunk 9]
```diff
@@ -20,7 +19,7 @@ function addHoverEvent(node: VisualElement<Element>, isActive: boolean) {
         }
 
         if (props[callbackName]) {
-            frame.update(() => props[callbackName](event, info))
+            props[callbackName](event, info)
```
Comment (adamseckel) [inline packages/framer-motion/src/gestures/hover.ts:22]: Is this an unrelated/drive-by change to no longer defer event handlers?

Comment (mattgperry) [reply packages/framer-motion/src/gestures/hover.ts:22]: It's actually entwined because without doing this we can add a two-frame delay between interactions and paint (which is counter to the point of this) as we have the frame for the event then the frame to resolve keyframes

Comment (adamseckel) [reply packages/framer-motion/src/gestures/hover.ts:22]: Ah!

**Review by adamseckel — COMMENTED** [review]
I wonder if before merging this, it would be worthwhile to create a FramerStudio branch with this alpha, and see what tests pass/fail as an indicator of if we are missing anything critical?

File: packages/framer-motion/src/render/dom/DOMKeyframesResolver.ts [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+
+        unresolvedKeyframes[unresolvedKeyframes.length - 1] = positionalValues[
+            name
+        ](
+            element.measureViewportBox(),
+            window.getComputedStyle(element.current)
+        ) as any
+
+        if (name === "height" && this.suspendedScrollY !== undefined) {
+            window.scrollTo(0, this.suspendedScrollY)
```
Comment (adamseckel) [inline packages/framer-motion/src/render/dom/DOMKeyframesResolver.ts:169]: Why do we need to scroll the window to 0?

**Review by adamseckel — APPROVED** [review]

File: packages/framer-motion/src/render/dom/DOMKeyframesResolver.ts (hunk 10 above)
Comment (mattgperry) [reply packages/framer-motion/src/render/dom/DOMKeyframesResolver.ts:169]: We're scrolling it to the value it was before measurement - the measurement can fuck with page scroll
